### PR TITLE
Tidy up into Cargo workspace & refactor for easier testing (v1.2.0)

### PR DESCRIPTION
**Pull request by Krulknul**
## Cargo workspace

I decided to create a cargo workspace out of this repo and split up the examples from the `radix-event-stream` crate. Cargo allows examples to live inside of crates so you can run them using `cargo run --example <name>`, but it is a bit limited when it comes to including a bit more library code as supplement to the examples (like some event declarations). Also, dependencies need to be added to dev-dependencies. Some of the examples are a bit more complex, so I decided it might be better to split that up into its own crate, with binary targets. I did just that.

## Refactoring for easier testing

I refactored most of the transaction processing code out of the `TransactionStreamProcessor` into a new `TransactionProcessor` struct. It handles processing single transactions, and also uses that to process a vec or slice of transactions. It pretty much does most of what `TransactionStreamProcessor` used to do, but without a `TransactionStream`, allowing you to just pass in the transactions. Then, `TransactionStreamProcessor` uses an instance of `TransactionProcessor` with a `TransactionStream` to process transactions from a stream

### Before:
Responsibilty of `TransactionStreamProcessor`:
* used by the user
* ties everything together
* gets transactions from a stream and calls transaction handlers on them
* inspects return values and handles retries
* calls logging hooks

### After:
Responsibilty of `TransactionStreamProcessor`:
* used by the user
* ties everything together
* gets transactions from a stream and gives them to `TransactionProcessor` for processing

Responsibilty of `TransactionProcessor`:
* calls transaction handlers on passed in transactions
* inspects return values and handles retries
* calls logging hooks

This allows you to skip using a `TransactionStream` and use the `TransactionProcessor` directly, passing in transactions, having it give back control when it's done.

At first, I implemented functions called `process_n` and `process_until_empty` on `TransactionStreamProcessor`, but I found that the experience for testing could be improved by refactoring the processing logic out and bypassing the channel stream. I also suspect it will become easier to test the processing logic down the line.

See https://github.com/ociswap/radix-event-stream/blob/workspace/crates/examples/src/bin/batching.rs for an example of a test using `TransactionProcessor` directly.

Additionally, I changed the version number to 1.2.0

### CODE REVIEW

File: crates/examples/src/bin/batching.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        let transaction = next.recv().await.unwrap();
+        transactions.push(transaction);
+    }
+
+    let mut processor =
+        TransactionProcessor::new(handler_registry, State { number: 1 });
+
+    // Process the transactions by passing in the transactions to the TransactionProcessor
+    // see also the process_transaction method to process only one transaction
+    processor.process_transactions(&transactions).await.unwrap();
```
Comment (fpieper) [inline crates/examples/src/bin/batching.rs]: This looks great and is exactly what we need to write the tests in the backend 👍

File: Cargo.toml [hunk 2]
```diff
@@ -1,51 +1,21 @@
-[package]
-name = "radix-event-stream"
-version = "1.1.1"
-edition = "2021"
+[workspace]
+members = [
+    "crates/radix-event-stream",
```
Comment (fpieper) [inline Cargo.toml]: I think it is better to not have them in the folder crates, but just flat in the directory.

File: Cargo.toml [hunk 3]
```diff
@@ -1,51 +1,21 @@
...
-gateway = ["radix-client"]
-file = ["serde_yaml"]
-channel = []
-
-
-[dev-dependencies]
-sqlx = { version = "0.7.4", features = ["sqlite"] }
-env_logger = "0.11.3"
+# Temporary fix until new gateway is released
+sbor-json = { git = 'https://github.com/Krulknul/radix-engine-toolkit.git', branch = "temporary-fix-bottlenose" }
```
Comment (fpieper) [inline Cargo.toml:62]: Guess would have also been fine to just add it to this repo, but this is fine too 👍

**Review by fpieper — COMMENTED** [review]
Thank you, looking really nice 👍 . Maybe just remove the nesting with the folder `crates`

File: Cargo.toml (hunk 3 above)
Comment (Krulknul) [reply Cargo.toml:62]: I expect them to release new gateway pretty soon and then we can just use radixdlt's repo

**Comment by fpieper** [comment]
@Krulknul looks good now thx 👍 

One last thing, I think I would use underscores for crate names instead of hyphens. Is probably more consistent. Read somewhere that it is also kind of recommended.

**Comment by Krulknul** [comment]
> @Krulknul looks good now thx 👍
> 
> One last thing, I think I would use underscores for crate names instead of hyphens. Is probably more consistent. Read somewhere that it is also kind of recommended.

I think I would be against that for now, since the name of the create was already radix-event-stream, and also the fact that the repo is named that way. Looked around online and couldn't really find a clear consensus on which is better.

**Comment by fpieper** [comment]
@Krulknul yes the repo is named that way and would keep that. However, Rust automatically converts hyphens to underscore in crate names if you are importing them. So that’s why I think it makes more sense to use underscore in the crate name

**Comment by Krulknul** [comment]
> @Krulknul yes the repo is named that way and would keep that. However, Rust automatically converts hyphens to underscore in crate names if you are importing them. So that’s why I think it makes more sense to use underscore in the crate name

Alright, I don't really have a strong opinion about it but I changed it now

**Review by fpieper — APPROVED** [review]
Thank you very much. Looking good to me now.

You can merge and release a new version 👍

**Comment by Krulknul** [comment]
@fpieper Thanks!

**Pull request merged**